### PR TITLE
Adding Resource Filters

### PR DESCRIPTION
**Pull request by rynowak**
This is a new filter stage that surrounds the existing model binding,
action, and result parts of the pipeline. Resource Filters run after
Authorization Filters.

The other major change is to support one of the primary scenarios for
Resource Filters. We want a filter to be able to modify the inputs the
model binding (formatters, model binders, value providers, etc) -- this
means that those changes need to be held on a context object and preserved
so that they can be used in the controller.

So, IActionBindingContextProvider is removed - the ActionBindingContext
will be created by the invoker. For now it will be part of the action
context.

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs [hunk 1]
```diff
@@ -146,6 +152,149 @@ private IFilter[] GetFilters()
...
         }
 
+        private async Task InvokeAllResourceFiltersAsync()
+        {
+            _cursor.SetStage(FilterStage.ResourceFilters);
+
+            var context = new ResourceExecutingContext(ActionContext, _filters);
+
+            context.InputFormatters = new List<IInputFormatter>(_inputFormatterProvider.InputFormatters);
+            context.InputFormatterSelector = _inputFormatterSelector;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs]: I'm not really sure what the design intent of input formatter selector is. Let's discuss when you're back.

File: src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs [hunk 2]
```diff
@@ -219,7 +368,36 @@ private IFilter[] GetFilters()
...
+            var bindingContext = new ActionBindingContext()
+            {
+                InputFormatters = _resourceExecutingContext.InputFormatters,
+                InputFormatterSelector = _resourceExecutingContext.InputFormatterSelector,
+                MetadataProvider = _modelMetadataProvider,
+                ModelBinder = new CompositeModelBinder(_resourceExecutingContext.ModelBinders),
+                ValidatorProvider = new CompositeModelValidatorProvider(_resourceExecutingContext.ValidatorProviders),
+            };
+
+            var valueProviders = new List<IValueProvider>();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs]: Added a static factory method for this. Due to the layer of MVC/ModelBinding, you still have to create a VPFContext.

File: src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs [hunk 3]
```diff
@@ -219,7 +368,36 @@ private IFilter[] GetFilters()
         private async Task InvokeAllActionFiltersAsync()
         {
             _cursor.SetStage(FilterStage.ActionFilters);
+
+            Debug.Assert(_resourceExecutingContext != null);
+            var bindingContext = new ActionBindingContext()
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs]: We could just put the ABC on the ResourceFilterContext, but that has some subtle differences.  I really don't think the copying is bad, it's 5-6 properties, and that's kinda what DTOs do. Let's discuss.

File: src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        /// <summary>
+        /// Executes the resource filter.
+        /// </summary>
+        /// <param name="context">The <see cref="ResourceExecutingContext"/>.</param>
+        /// <param name="next">
+        /// The <see cref="ResourceExecutionDelegate"/>. Invoked to execute the next
+        /// resource filter, or the remainder of the pipeline.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Task"/> which will complete when the remainer of the pipeline completes.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs]: fixed

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 5]
```diff
@@ -887,23 +885,24 @@ public virtual Task<bool> TryUpdateModelAsync<TModel>([NotNull] TModel model)
             [NotNull] params Expression<Func<TModel, object>>[] includeExpressions)
            where TModel : class
         {
-            if (BindingContextProvider == null)
+            if (ActionContext.BindingContext == null)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: wondering if these provide any value ```TryUpdateModel``` would be called from within an action by which time this would have been initialized.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 6]
```diff
@@ -824,15 +821,15 @@ public virtual Task<bool> TryUpdateModelAsync<TModel>([NotNull] TModel model)
...
             where TModel : class
         {
-            if (BindingContextProvider == null)
+            if (ActionContext.BindingContext == null)
             {
-                var message = Resources.FormatPropertyOfTypeCannotBeNull(nameof(BindingContextProvider), 
-                                                                         GetType().FullName);
+                var message = Resources.FormatPropertyOfTypeCannotBeNull(
+                    nameof(ActionContext.BindingContext),
+                    typeof(ActionContext).Name);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: FullName?

File: src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs]: I don't think the selector should be changeable by the user (at least not via a ResourceFilter). 
```InputFormatterSelector``` does on the input side what ```ObjectResult``` does on the output side. It selects the input formatter by calling ```CanRead``` on input formatter.

File: src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs (hunk 3 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs]: Things which we would not want a ResourceFilter to touch but are needed in binding would make sense only in ABC and not in ResourceBindingContext (For ex InputFormatterSelector).

File: src/Microsoft.AspNet.Mvc.WebApiCompatShim/ApiController.cs [hunk 7]
```diff
@@ -397,15 +397,16 @@ public void Validate<TEntity>(TEntity entity)
...
         {
+            var bindingContext = ActionContext.BindingContext;
+            var modelMetadata = bindingContext.MetadataProvider.GetMetadataForType(() => entity, typeof(TEntity));
+
             var bodyValidationExcludeFiltersProvider = Context.RequestServices
                                                               .GetRequiredService<IValidationExcludeFiltersProvider>();
             var validator = Context.RequestServices.GetRequiredService<IBodyModelValidator>();
-            var metadataProvider = Context.RequestServices.GetRequiredService<IModelMetadataProvider>();
-            var modelMetadata = metadataProvider.GetMetadataForType(() => entity, typeof(TEntity));
-            var validatorProvider = Context.RequestServices.GetRequiredService<ICompositeModelValidatorProvider>();
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.WebApiCompatShim/ApiController.cs]: nice...!

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs [hunk 8]
```diff
@@ -1203,6 +1203,647 @@ private struct SampleStruct
...
+            var resourceFilter = new Mock<IAsyncResourceFilter>(MockBehavior.Strict);
+            resourceFilter
+                .Setup(f => f.OnResourceExecutionAsync(It.IsAny<ResourceExecutingContext>(), It.IsAny<ResourceExecutionDelegate>()))
+                .Returns<ResourceExecutingContext, ResourceExecutionDelegate>(async (c, next) =>
+                {
+                    context = await next();
+                })
+                .Verifiable();
+
+            var invoker = CreateInvoker(new IFilter[] { resourceFilter.Object });
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: nit: do not need to initialize an array here.

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs [hunk 9]
```diff
@@ -1203,6 +1203,647 @@ private struct SampleStruct
...
+            var resourceFilter2 = new Mock<IAsyncResourceFilter>(MockBehavior.Strict);
+            resourceFilter2
+                .Setup(f => f.OnResourceExecutionAsync(It.IsAny<ResourceExecutingContext>(), It.IsAny<ResourceExecutionDelegate>()))
+                .Returns<ResourceExecutingContext, ResourceExecutionDelegate>((c, next) =>
+                {
+                    throw expected;
+                })
+                .Verifiable();
+
+            var invoker = CreateInvoker(new IFilter[] { resourceFilter1.Object, resourceFilter2.Object }, actionThrows: true);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: Are we keeping actionThrows true intentionally?

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: Doesn't much matter here. I think generally these tests have the action throw when we don't expect it to run.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: sure

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs (hunk 8 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: sure, there's plenty more like this though, and it doesn't seem useful to try and clean them as it's just sugar in the API.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: This is just in case you do anything wonky. We don't expect users to see this in practice unless they are far off the beaten path.

File: src/Microsoft.AspNet.Mvc.Core/ActionContext.cs [hunk 10]
```diff
@@ -40,11 +41,16 @@ public ActionContext([NotNull] RouteContext routeContext, [NotNull] ActionDescri
 
         public ActionDescriptor ActionDescriptor { get; private set; }
 
+
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ActionContext.cs]: remove.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 11]
```diff
@@ -796,7 +793,7 @@ public virtual void OnActionExecuted([NotNull] ActionExecutedContext context)
         }
 
         /// <summary>
-        /// Updates the specified <paramref name="model"/> instance using values from the controller's current
+        /// Updates the specified <paramref name="model"/> instance using values from the controller's current
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: nit: remove whitespace at the end.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 12]
```diff
@@ -850,33 +847,34 @@ public virtual Task<bool> TryUpdateModelAsync<TModel>([NotNull] TModel model)
...
+                    nameof(ActionContext.BindingContext),
+                    typeof(ActionContext).FullName);
                 throw new InvalidOperationException(message);
             }
 
-            var bindingContext = await BindingContextProvider.GetActionBindingContextAsync(ActionContext);
-            return await ModelBindingHelper.TryUpdateModelAsync(model,
+            return await ModelBindingHelper.TryUpdateModelAsync(
+                model,
                                                                 prefix,
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: formatting.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 13]
```diff
@@ -886,27 +884,28 @@ public virtual Task<bool> TryUpdateModelAsync<TModel>([NotNull] TModel model)
...
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: formatting.

File: src/Microsoft.AspNet.Mvc.Core/ControllerActionInvokerProvider.cs [hunk 14]
```diff
@@ -2,26 +2,43 @@
...
+            IControllerActionArgumentBinder argumentBinder,
+            IModelMetadataProvider modelMetadataProvider,
+            IInputFormatterSelector inputFormatterSelector,
+            IModelBinderProvider modelBinderProvider,
+            IModelValidatorProviderProvider modelValidationProviderProvider,
+            IValueProviderFactoryProvider valueProviderFactoryProvider)
         {
             _controllerFactory = controllerFactory;
-            _inputFormattersProvider = inputFormattersProvider;
+            _inputFormatterProvider = inputFormatterProvider;
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ControllerActionInvokerProvider.cs]: why is the name changed? the interface is ```IInputFormattersProvider ```

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ControllerActionInvokerProvider.cs]: The interface is wrong, I'll rename it.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ControllerActionInvokerProvider.cs]: Actually there's a bunch of these that are wrong. I'm going to rename the field and just leave it.

File: src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// A filter which surrounds execution of model binding, the action (and filters) and the action result
+    /// (and filters).
+    /// </summary>
+    public interface IAsyncResourceFilter : IFilter
+    {
+        /// <summary>
+        /// Executes the resource filter.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs]: be consistent in the descrption of async and non async versions.

File: src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        /// </summary>
+        /// <param name="context">The <see cref="ResourceExecutingContext"/>.</param>
+        /// <param name="next">
+        /// The <see cref="ResourceExecutionDelegate"/>. Invoked to execute the next
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs]: ```invoked```

File: src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs (hunk 15 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs]: Not sure what you mean?

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+            set
+            {
+                _exception = null;
+                _exceptionDispatchInfo = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether not the current <see cref="Exception"/> has been handled.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs]: ```whether or not```

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+{
+    /// <summary>
+    /// A context for resource filters.
+    /// </summary>
+    public class ResourceExecutedContext : FilterContext
+    {
+        private Exception _exception;
+        private ExceptionDispatchInfo _exceptionDispatchInfo;
+
+        public ResourceExecutedContext(ActionContext actionContext, IList<IFilter> filters)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs]: docs?

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+            : base(actionContext, filters)
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets a value which indicates whether or execution was canceled by a resource filter.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs]: ```whether or not```

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutingContext.cs [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// A context for resource filters. Allows modification of services and values used for
+    /// model binding.
+    /// </summary>
+    public class ResourceExecutingContext : FilterContext
+    {
+        public ResourceExecutingContext(ActionContext actionContext, IList<IFilter> filters)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutingContext.cs]: docs

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutionDelegate.cs [hunk 21]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public delegate Task<ResourceExecutedContext> ResourceExecutionDelegate();
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutionDelegate.cs]: docs?

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs [hunk 22]
```diff
@@ -1203,6 +1203,647 @@ private struct SampleStruct
...
+
+            var resourceFilter3 = new Mock<IAsyncResourceFilter>(MockBehavior.Strict);
+            var actionFilter = new Mock<IAsyncActionFilter>(MockBehavior.Strict);
+            var resultFilter = new Mock<IAsyncResultFilter>(MockBehavior.Strict);
+
+            var invoker = CreateInvoker(
+                new IFilter[]
+                {
+                    resourceFilter1.Object, // This filter should see the result retured from resourceFilter2
+                    resourceFilter2.Object,
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: seems odd to miss out a comment for this one :smile:  ```// This filter short circuits.```

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs [hunk 23]
```diff
@@ -1203,6 +1203,647 @@ private struct SampleStruct
...
+                    resourceFilter3.Object, // This shouldn't run - it will throw if it does
+                    actionFilter.Object, // This shouldn't run - it will throw if it does
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: why not an exception filter as well ?

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs [hunk 24]
```diff
@@ -1203,6 +1203,647 @@ private struct SampleStruct
...
+
+            // Act
+            var exception = await Assert.ThrowsAsync<InvalidOperationException>(invoker.InvokeAsync);
+
+            // Assert
+            Assert.Equal(message, exception.Message);
+        }
+
+        [Fact]
+        public async Task InvokeAction_AuthorizationFilter_ChallengePreventsResourceFiltersFromRunning()
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: probably one more where an authorization filter throws?

File: test/Microsoft.AspNet.Mvc.Core.Test/InputObjectBindingTests.cs [hunk 25]
```diff
@@ -1,103 +1,13 @@
 ﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/InputObjectBindingTests.cs]: do we need this class anymore?

File: src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs (hunk 15 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs]: nvm... my bad.

**Comment by harshgMSFT** [comment]
looks pretty close to done :watch: on doc cleanup.

File: src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/FilterActionInvoker.cs]: So the change here is that `ActionBindingContext` is now owned by `IContextAccessor` and it no longer has properties that refer to services - so model metadata provider and input formatter selector are gone.

File: src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs (hunk 16 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs]: It's the beginning of a sentence

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs (hunk 17 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs]: fixed

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs (hunk 18 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs]: added

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs (hunk 19 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutedContext.cs]: fixed

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutingContext.cs (hunk 20 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutingContext.cs]: Fixed

File: src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutionDelegate.cs (hunk 21 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/ResourceExecutionDelegate.cs]: fixed

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs (hunk 23 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: sure

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs (hunk 24 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: Added a resource filter to InvokeAction_ExceptionInAuthorizationFilterCannotBeHandledByOtherFilters

File: test/Microsoft.AspNet.Mvc.Core.Test/InputObjectBindingTests.cs (hunk 25 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/InputObjectBindingTests.cs]: Had to copy some of the models around. The file is delorted now.

**Comment by rynowak** [comment]
@harshgMSFT updated

File: test/WebSites/PrecompilationWebSite/Views/Home/_ViewStart.cshtml [hunk 26]
```diff
@@ -1,4 +1,4 @@
 @using System.Reflection
 @{ Layout = "/views/Home/Layout.cshtml";}
 _viewstart:@GetType().GetTypeInfo().Assembly.FullName
- 
\ No newline at end of file
+
```
Comment (harshgMSFT) [inline test/WebSites/PrecompilationWebSite/Views/Home/_ViewStart.cshtml]: unrelated?

File: test/WebSites/PrecompilationWebSite/Views/Home/Layout.cshtml [hunk 27]
```diff
@@ -1,2 +1,2 @@
 Layout:@GetType().GetTypeInfo().Assembly.FullName
-@RenderBody() 
\ No newline at end of file
+@RenderBody()
```
Comment (harshgMSFT) [inline test/WebSites/PrecompilationWebSite/Views/Home/Layout.cshtml]: unrelated?

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs (hunk 23 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: is this fixed?

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 28]
```diff
@@ -850,33 +853,34 @@ public virtual Task<bool> TryUpdateModelAsync<TModel>([NotNull] TModel model)
...
-        /// Updates the specified <paramref name="model"/> instance using values from the controller's current
+        /// Updates the specified <paramref name="model"/> instance using values from the controller's current 
         /// <see cref="IValueProvider"/> and a <paramref name="prefix"/>.
         /// </summary>
         /// <typeparam name="TModel">The type of the model object.</typeparam>
         /// <param name="model">The model instance to update.</param>
         /// <param name="prefix">The prefix to use when looking up values in the current <see cref="IValueProvider"/>.
         /// </param>
-        /// <param name="includeExpressions"> <see cref="Expression"/>(s) which represent top-level properties
+        /// <param name="includeExpressions"> <see cref="Expression"/>(s) which represent top-level properties
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: white space here and many such places in this file.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs [hunk 29]
```diff
@@ -886,27 +890,28 @@ public virtual Task<bool> TryUpdateModelAsync<TModel>([NotNull] TModel model)
...
-                                                                prefix,
-                                                                ActionContext.HttpContext,
-                                                                ModelState,
-                                                                bindingContext.MetadataProvider,
-                                                                bindingContext.ModelBinder,
-                                                                bindingContext.ValueProvider,
-                                                                bindingContext.ValidatorProvider,
-                                                                includeExpressions);
+            return await ModelBindingHelper.TryUpdateModelAsync(
+                model,
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: nit: whitespace

File: src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs (hunk 16 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IAsyncResourceFilter.cs]: my bad misread it.

**Comment by harshgMSFT** [comment]
:shipit: after reverting the unrelated changes.

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs (hunk 23 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: yes. it's below the fold

File: test/WebSites/PrecompilationWebSite/Views/Home/_ViewStart.cshtml (hunk 26 above)
Comment (rynowak) [inline test/WebSites/PrecompilationWebSite/Views/Home/_ViewStart.cshtml]: the tests do this. this is why IO in tests is evil. I'll revert it.

File: test/WebSites/PrecompilationWebSite/Views/Home/Layout.cshtml (hunk 27 above)
Comment (rynowak) [inline test/WebSites/PrecompilationWebSite/Views/Home/Layout.cshtml]: the tests do this. this is why IO in tests is evil. I'll revert it.

File: src/Microsoft.AspNet.Mvc.Core/Controller.cs (hunk 28 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controller.cs]: not going to bother, this is something we can and do automate
